torch_sim/models/metatensor.py

```diff
@@ -0,0 +1,328 @@
+                and other system information. Can be either a SimState object or a
+                dictionary with the relevant fields.
+
+        Returns:
+            dict[str, torch.Tensor]: Dictionary containing:
+                - 'energy': System energies with shape [n_systems]
+                - 'forces': Atomic forces with shape [n_atoms, 3] if compute_forces=True
+                - 'stress': System stresses with shape [n_systems, 3, 3] if
+                    compute_stress=True
+        """
+        # Extract required data from input
+        if isinstance(state, dict):
+            state = SimState(**state, masses=torch.ones_like(state["positions"]))
+
+        # Input validation is already done inside the forward method of the
+        # MetatensorAtomisticModel class, so we don't need to do it again here.
+
+        atomic_numbers = state.atomic_numbers
+        cell = state.row_vector_cell
+        positions = state.positions
+        pbc = state.pbc
+
+        # Check dtype (metatensor models require a specific input dtype)
+        if positions.dtype != self._dtype:
+            raise TypeError(
+                f"Positions dtype {positions.dtype} does not match model dtype "
+                f"{self._dtype}"
+            )
+
+        # Compared to other models, metatensor models have two peculiarities:
+        # - different structures are fed to the models separately as a list of System
+        #   objects, and not as a single graph-like batch
+        # - the model does not compute forces and stresses itself, but rather the
+        #   caller code needs to call torch.autograd.grad or similar to compute them
+        #   from the energy output
+
+        # Process each system separately
+        systems: list[System] = []
+        strains = []
+        for b in range(len(cell)):
+            system_mask = state.batch == b
+            system_positions = positions[system_mask]
+            system_cell = cell[b]
+            system_pbc = torch.tensor(
+                [pbc, pbc, pbc], device=self._device, dtype=torch.bool
+            )
+            system_atomic_numbers = atomic_numbers[system_mask]
+
+            # Create a System object for this system
+            if self._compute_forces:
+                system_positions.requires_grad_()
+            if self._compute_stress:
+                strain = torch.eye(
+                    3, device=self._device, dtype=self._dtype, requires_grad=True
+                )
+                system_positions = system_positions @ strain
+                system_cell = system_cell @ strain
+            system = System(
+                positions=system_positions,
+                types=system_atomic_numbers,
+                cell=system_cell,
+                pbc=system_pbc,
+            )
+
+            # Calculate neighbor list(s) for this system
+            for neighbor_list_options in self._requested_neighbor_lists:
```

Comment:
Do all models in metatensor use a radial cutoff for the neighbor list?

If not, consider setting a dynamic `memory_scales_with` @propery. The default memory_scales_with from ModelInterface will return "n_atoms_x_density", which is only correct for models with radial cutoffs.

For max-neighbors based cutoffs, the memory scaler will be incorrect and the autobatching will subtly break. See [here](https://radical-ai.github.io/torch-sim/tutorials/autobatching_tutorial.html#Understanding-Memory-Requirements) for a more detailed discussion.

Reply:
Yes, at the moment this interface only supports a fixed cutoff radius, although metatensor itself would allow a clamped number of neighbors by simply feeding a filtered neighbor list. However, we have never tested this possibility in training or in production, and therefore I would keep it as is for now

Reply:
There is also the possibility for model to set `interaction_range` to infinity, which is used for long-range models. I don't know if these would work with the current code, I'll check.

Reply:
The model scaling can be handled manually for edge cases, so I am happy leaving it as `n_atoms_x_density`